### ISSUE TITLE
5.0 imcompatible changes plan

### ISSUE THREAD

**Issue by cben** [abonas/kubeclient#435]
- [x] changed `under_score` algorithm https://github.com/abonas/kubeclient/pull/355
- [ ] change watch exceptions behavior https://github.com/abonas/kubeclient/pull/275#issuecomment-591928097
- [ ] drop EOL ruby versions
- [ ] I'm thinking about making all methods not documented in README `private`?
  Specifically `get_entities`, `patch_entity` etc?

**Comment by cben** [comment]
- What about Remove deprecated `KubeException` #293?
  => Not now. As explained there, `HttpError` is not sufficient, we need some top-level exception.  Should probably introduce a `Kubeclient::Error` but that's a compatible change, and should give users time to adopt it before dropping `KubeException`.

**Comment by cben** [comment]
Apologies for delays.  Working from home with kids has been... challenging.
I hope to get a lot done next week.

**Comment by grosser** [comment]
I'd propose: 
- get rid of most of the api:
User has to set the api version and resource anyway, so much more efficient and simple to call `client.get("nodes")` without doing the extra api discovery 🤷 
- remove struct and return hashes ... it's an expensive crutch :D

**Comment by cben** [comment]
> get rid of most of the api

That will be an entirely different gem :smile:
There is a non-breaking road though: first implement #332, then much later see how well people adopted new style and how people feel about dropping existing API.

> remove struct and return hashes

Now that it's simple matter of ``as: :parsed` / `as: :parsed_symbolized`, I don't see a large enough payoff to justify breaking compatibility.
We could do small nudges like make `as:` required in `Client` contructor to stop encouraging a bad choice.  Or even make people require and inject recursive-open-struct and stop depending on it?
But _forcing_ people with lots of working `.foo.bar` code to rewrite everything to `[:foo][:bar]` form would be a disservice IMHO.

If you care about this, start by changing README to prominently explain why the new modes are recommended.

In short, both of these points sound more "8.0" (and maybe never) then "5.0" to me :)

**Comment by cben** [comment]
- [ ] `.rest_client` is definitely going private (or away) to unblock #466.

**Comment by grosser** [comment]
#332 sounds like a good step forward
... requiring `as` also sounds good (with or without removing struct
runtime dependency)

**Comment by grosser** [comment]
stumbled into the whole url parsing mess again 😞 
so add to the list:
- make version required
- make version `apps/v1` instead of mashing up host/apis/apps and v1
- dynamically set `api` or `apis` depending on version == v1
... and remove all the old glue/parsing code to have a single interface

**Comment by grosser** [comment]
plz start a v5 branch so I can make PRs against that

**Comment by cben** [comment]
master is already v5, recent releases were from backports to v4.y

**Comment by cben** [comment]
Paternity leave ended :baby:, sleeplessness continues, I didn't make progress I hoped for :-(

- If someone wants to help, one simple but important step is update changelog with everything unreleased we've merged on master.
